### ISSUE TITLE
Cannot reproduce Apollo Client 2 cache-and-network fetch policy 

### ISSUE THREAD

**Issue by olivierlauzon85** [apollographql/apollo-client#11628]
Hello 👋

As we're transitioning from Apollo Client 2 to 3, we've encountered significant blockers due to changes in the cache-and-network fetch policy. Specifically, the new version initiates network requests with any cache change, leading to an unsustainable number of requests for our large-scale application.

We attempted to follow the recommended workaround to mimic Apollo Client 2's behavior by setting nextFetchPolicy to "cache-first" (both on individual queries, or on the config like suggested in https://github.com/apollographql/apollo-client/issues/6760#issuecomment-668188727). Unfortunately, this does not restore the previous functionality.

Mutations that update the cache, like changing a profile picture, now result in additional network fetches, despite attempts to use the suggested nextFetchPolicy workarounds. The picture do get updated in the UI, but there are way too many extra network requests going out.

Another major problem we have is on a pubsub that manages unread notifications. Whenever there is a change on unread, the subscription takes care of fetching the new data so whoever's subscribed can just use that new data. In Apollo Client 2 this would work well: clientA subcribes, whenever clientB makes a relevant change the subscription would query the relevant data and update the cache. No network request would happen on clientA when he receives the subscription update. On Apollo Client 3, even with the suggested fix to have the same behaviour, when clientA receives the subscription update, a network request is made to re-fetch the data that the subscription just took care of. This makes way too many useless network calls.

TLDR; Apollo Client 3 makes extra network requests with cache-and-network policy, even with cache-first as nextFetchPolicy. We'd like to hear about any potential reasons to why the suggested fix doesn't bring back the Apollo Client 2 behaviour.

Thanks!

**Comment by jeel2308** [comment]
Hi, which apollo client version are you using, currently?

**Comment by phryneas** [comment]
Hi Olivier,

you are right that our recommendation to reproduce this old behaviour is to start with `cache-and-network` and then use a `nextFetchPolicy` of `cache-first`.

Your query will immediately move over to the new `fetchPolicy` once the request is started, so that gets me thinking that this might not be your problem here.

There are situations where even with a `fetchPolicy` of `cache-first`, a mutation triggers an outgoing query. Namely, if the mutation (or another query) overwrites a part of the cache with data that is not sufficiently complete to answer your initial query.

That would usually happen if your request and mutation somehow overlap, and for some reason your cache cannot be normalized.

When you inspect the cache in the devtools, is is normalized properly? If not, you might be missing `id` fields in your queries in some places 🤔

**Comment by olivierlauzon85** [comment]
Hi,

Ok that clears things up a bit. Makes me think that the change in behaviour is likely caused by the change in [InMemoryCache merge](https://github.com/apollographql/apollo-client/pull/5603), but this will likely require a lot more investigation on our side because we were relying on that mechanism in a lot of places. I believe this will prevent us from upgrading to Apollo Client 3 in the foreseeable future.

Thanks!

**Comment by jeel2308** [comment]
Hi, @olivierlauzon85 
I have done this exercise already. It is tedious and you will definitely encounter more issues. Apollo client v3 is good but it also introduces new bugs and break things which were working fine in apollo client v2. So, If you are seriously planning this, then I will suggest to read `CHANGELOG.md` carefully to understand implications.

**Comment by jerelmiller** [comment]
I'm doing a bit of housekeeping and I believe the original question might be answered. I'm going to go ahead and close this issue. If you still need assistance, please let me know and I'd be happy to open this back up. Thanks!
